proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+
+There could be a need to support various types of the scripts in the same environment. E.g. in a web framework there 
+could be routing scripts and content scripts. For such cases one more component should be configured:
+- *Selector* - the service that recognizes a script using file name or text
+
+The same component could be used for script type selection in an IDE.
+
+#### Caching
+
+Since calling Kotlin compiler could be quite a heavy operation in comparison with typical script execution, the caching
+of the compiled script should be supported by the compilation platform.
+
+#### Execution lifecycle
+
+The script is executed in the following pipeline:
+- selection - the *Selectors* are called on the script to choose the set of services that will handle it
+- preprocessing - the *Preprocessor* takes the script and the environment and extracts the compilation configuration
+- compilation - the *Platform* takes the script, and the compilation configuration and provides the compiled class
+- instantiation - the *Executor* takes the compiled class, the environment, and the compilation configuration and
+  instantiates the script object, passing the required parameters from the environment to the constructor
+- execution - the *Executor* takes the instantiated script and executes the appropriate method, passing arguments
+  from the environment to it; this step could be repeated many times
+
+#### Processing in an IDE
+
+The script is handled in an IDE in the following sequence (*Note that the first two steps are the same as in the 
+execution pipeline*):
+- selection - the *Selectors* takes the script and selects the set of services that will handle it
+- preprocessing - the *Preprocessor* takes the script and the environment and extracts the compilation configuration
+- resolution info extraction - the *Platform* takes the script and the compilation configuration and provides the
+  info required for rich editing the script
+
+### Standalone scripting
+
+Standalone scripting applications include command-line utilities and a standalone Kotlin REPL.
+
+Standalone scripting is a variant of the embedded scripting with hosts provided with the Kotlin distribution.
+
+#### Hosts
+
+The standalone script could be executed e.g. using command line Kotlin compiler:
+
+`kotlinc -script myscript.kts`
+
+Or with some hypothetical dedicated runner, that we might want to include into the distribution:
+
+`kotlinsh myscrtipt.ktsh`
+
+To be able to use the Kotlin scripts in a Unix shell environment, the *shebang* (`#!`) syntax should be supported 
+at the beginning of the script:
+```
+#! /path/to/kotlin/script/runner -some -params
+```
+
+#### Parameters
+
+For the command line usage the support for script parameters is needed. The simplest form is to assume that the script
+has access to the `args: Array<String>` property/parameter. More advanced is to support a declaration of the typed 
+parameters, e.g.:
+
+```
+@file:param("name", String::class)
```

Comment:
With the `::class` literal, you can only reference a class, not a type. For example, this way you can't specify a parameter with the type `String?`, or `List<String>`.